src/flb_utils.c

```diff
@@ -735,48 +743,66 @@ int flb_utils_write_str(char *buf, int *off, size_t size,
             i += (hex_bytes - 1);
         }
         else if (c > 0xFFFF) {
-            hex_bytes = flb_utf8_len(str + i);
+            utf_sequence_length = flb_utf8_len(str + i);
             if (available - written < 6) {
                 return FLB_FALSE;
             }
 
-            if (i + hex_bytes > str_len) {
+            if (i + utf_sequence_length > str_len) {
                 break; /* skip truncated UTF-8 */
             }
 
             is_valid = FLB_TRUE;
-            for (b = 0; b < hex_bytes; b++) {
+            for (utf_sequence_number = 0; utf_sequence_number < utf_sequence_length;
+                utf_sequence_number++) {
                 /* Leading characters must start with bits 11 */
-                if (b == 0 && ((str[i] & 0xC0) != 0xC0)) {
+                if (utf_sequence_number == 0 && ((str[i] & 0xC0) != 0xC0)) {
                     /* Invalid unicode character. replace */
                     flb_debug("[pack] unexpected UTF-8 leading byte, "
                              "substituting character with replacement character");
+                    tmp[utf_sequence_number] = str[i];
                     ++i; /* Consume invalid leading byte */
+                    utf_sequence_length = utf_sequence_number + 1;
                     is_valid = FLB_FALSE;
                     break;
                 }
                 /* Trailing characters must start with bits 10 */
-                else if (b > 0 && ((str[i] & 0xC0) != 0x80)) {
+                else if (utf_sequence_number > 0 && ((str[i] & 0xC0) != 0x80)) {
                     /* Invalid unicode character. replace */
                     flb_debug("[pack] unexpected UTF-8 continuation byte, "
                              "substituting character with replacement character");
                     /* This byte, i, is the start of the next unicode character */
+                    utf_sequence_length = utf_sequence_number;
                     is_valid = FLB_FALSE;
                     break;
                 }
 
-                tmp[b] = str[i];
+                tmp[utf_sequence_number] = str[i];
                 ++i;
             }
             --i;
 
             if (is_valid) {
-                encoded_to_buf(p, tmp, hex_bytes);
-                p += hex_bytes;
+                encoded_to_buf(p, tmp, utf_sequence_length);
+                p += utf_sequence_length;
             }
             else {
-                encoded_to_buf(p, flb_utils_replacement_character, 3);
-                p += 3;
+                encoded_to_buf(p, flb_utils_replace_fragment_start,
```

Comment:
Shouldn't you calculate the length of the whole string you're injecting and verify if there's enough space available before doing anything in this path? Unless I'm wrong there are 34 characters not including the hex encoded bytes and spaces between them so it's substantially higher than what's ensured before.

Reply:
No, this function does not take out buffer size as a parameter, and so there is no way to verify the expanded size will fit in the out buffer. Please see the comment here: https://github.com/fluent/fluent-bit/pull/4297#issuecomment-969794729